version: '3.8'

services:
  namenode:
    build:
      context: ./namenode
    container_name: namenode
    restart: on-failure
    ports:
      - 9870:9870
      - 9000:9000
      - 80:9870
    volumes:
      - hadoop_namenode:/hadoop/dfs/name
    environment:
      - CLUSTER_NAME=test
    env_file:
      - ./hadoop.env

  datanode:
    build:
      context: ./datanode
    container_name: datanode
    hostname: datanode
    restart: on-failure
    ports:
      - 9864:9864
    volumes:
      - hadoop_datanode:/hadoop/dfs/data
    environment:
      SERVICE_PRECONDITION: "namenode:9870"
    env_file:
      - ./hadoop.env
    depends_on:
      - namenode

  resourcemanager:
    build:
      context: ./resourcemanager
    container_name: resourcemanager
    restart: on-failure
    ports:
      - 8088:8088
    environment:
      SERVICE_PRECONDITION: "namenode:9000 namenode:9870 datanode:9864"
    env_file:
      - ./hadoop.env
    depends_on:
      - namenode
      - datanode

  nodemanager:
    container_name: nodemanager
    build:
      context: ./nodemanager
    restart: on-failure
    environment:
      SERVICE_PRECONDITION: "namenode:9000 namenode:9870 datanode:9864 resourcemanager:8088"
    env_file:
      - ./hadoop.env
    depends_on:
      - namenode
      - datanode

  historyserver:
    build:
      context: ./historyserver
    container_name: historyserver
    restart: on-failure
    ports:
      - 8188:8188
    environment:
      SERVICE_PRECONDITION: "namenode:9000 namenode:9870 datanode:9864 resourcemanager:8088"
    volumes:
      - hadoop_historyserver:/hadoop/yarn/timeline
    env_file:
      - ./hadoop.env
    depends_on:
      - namenode
      - datanode

volumes:
  hadoop_namenode: null
  hadoop_datanode: null
  hadoop_historyserver: null
